lib/improver/nowcasting/optical_flow.py

```diff
@@ -321,23 +321,26 @@ def grid_spacing(coord):
         original_datetime, = \
             (cube.coord("time").units).num2date(cube.coord("time").points)
         new_datetime = original_datetime + timestep
-        new_time = (cube.coord("time").units).date2num(new_datetime)
 
-        if np.int64(new_time) == new_time:
-            new_time = np.int64(new_time)
-        else:
-            new_time = np.float64(new_time)
+        new_time = (cube.coord("time").units).date2num(new_datetime)
 
         advected_cube.coord("time").points = new_time
+        advected_cube.coord("time").convert_units(
+            "seconds since 1970-01-01 00:00:00")
+        advected_cube.coord("time").points = (
+            np.around(advected_cube.coord("time").points).astype(np.int64))
 
-        forecast_period_seconds = timestep.total_seconds()
-
-        forecast_period_seconds = np.float32(forecast_period_seconds)
+        advected_cube.coord("forecast_reference_time").convert_units(
+            "seconds since 1970-01-01 00:00:00")
+        frt_coord = cube.coord("forecast_reference_time")
+        advected_cube.coord("forecast_reference_time").points = (
+            frt_coord.points.astype(np.int64))
 
+        forecast_period_seconds = timestep.total_seconds()
+        forecast_period_seconds = np.int32(forecast_period_seconds)
         forecast_period_coord = AuxCoord(forecast_period_seconds,
                                          standard_name="forecast_period",
                                          units="s")
-
         try:
             advected_cube.remove_coord("forecast_period")
```

Comment:
Not related to your change but suggest this code block (lines 344-348) might be better as:
```
try:
    advected_cube.replace_coord(forecast_period_coord)
except CoordinateNotFoundError:
    advected_cube.add_aux_coord(forecast_period_coord)
```

Reply:
OK!  Seems like it's safer to leave it as it is anyway, then - I didn't realise things like units and datatypes all had to match.  Thanks for looking into it.